src/MySqlConnector/MySqlClient/Results/ResultSet.cs

```diff
@@ -0,0 +1,358 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Threading;
+using System.Threading.Tasks;
+using MySql.Data.Protocol.Serialization;
+using MySql.Data.Serialization;
+
+namespace MySql.Data.MySqlClient.Results
+{
+	internal class ResultSet
+	{
+		internal static readonly Task<bool> FalseTask = Task.FromResult(false);
+		internal static readonly Task<bool> TrueTask = Task.FromResult(true);
+
+		public ResultSet(MySqlDataReader dataReader)
+		{
+			DataReader = dataReader;
+		}
+
+		internal async Task<ResultSet> ReadResultSetHeaderAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
```

Comment:
Nitpick: I would consider this member to be part of the "public" API of the class (irrespective of the class's `internal` accessibility), following Eric's arguments here: https://ericlippert.com/2014/09/15/internal-or-public/

(Most of the methods on this type I would make `public`.)